### DISCUSSION TITLE
Spatial extension support

### DISCUSSION THREAD

**Discussion by VGSML** [duckdb/pg_duckdb#606] [Ideas]
Do you plan to implement DuckDB spatial extension support, mainly ST_Read?
Will it be PostGIS types support?

**Comment by wuputah** [comment]
Yes, we would like to see spatial / GIS support, it's just a matter of time and priorities. Determining how it interacts with postgis will certainly need to be part of the design.

**Comment by VGSML** [comment]
Thanks, will wait.

**Comment by philiporlando** [comment]
+1 for PostGIS support!

**Comment by chance-spatial** [comment]
Does it support spatial extension yet?
looks like functions are not registered. I have run `postgres=# select duckdb.install_extension('spatial');` first. But when I run `postgres=# SELECT ST_AsText(ST_MakeEnvelope(0,0,1,1));`
ERROR:  function st_makeenvelope(integer, integer, integer, integer) does not exist
LINE 1: SELECT ST_AsText(ST_MakeEnvelope(0,0,1,1));
                         ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.

**Comment by VGSML** [comment]
I think it won't work like your example. The main difficulty to implement is postgis—duckdb function mapping. In the PG should be used normal Postgis functions or duckdb functions from the schema duckdb. Postgis geometry is stored in the ewkb format, duckdb stores wkb - without srid and BBBox information, than could add additional difficulty.
Also there is no geography type in the duckdb spatial extension.

A lot of things to do ....

**Comment by Maxxen** [comment]
> I think it won't work like your example. The main difficulty to implement is postgis—duckdb function mapping. In the PG should be used normal Postgis functions or duckdb functions from the schema duckdb. Postgis geometry is stored in the ewkb format, duckdb stores wkb - without srid and BBBox information, than could add additional difficulty. Also there is no geography type in the duckdb spatial extension.
> 
> A lot of things to do ....

PostGIS doesn't store EWKB internally, and DuckDB doesn't store WKB internally either. Both use their own internal format and while they are not identical, the DuckDB format is very much inspired by PostGIS (both store BBOX:es). The major difference is that DuckDB does not store the SRID in the geometry value itself (and doesn't really have the concept of integer SRIDs in general).

I imagine the easiest solution is to just cast/convert to and from WKB when transferring geometries between DuckDB/PostGIS as ISO WKB basically models the common subset of spatial type support between the two systems.

**Reply by josephduchesne** [reply]
Is there any particular reason why SRID is omitted from the DuckDB geometry (or at least column definition?). It's a pretty fundamental part of the data, and only one additional (and optional) integer, since without it no other GIS system can really operate on the data without making assumptions the moment the source data has row or column specific SRIDs.

**Reply by Maxxen** [reply]
Hello! 

I don't think it's desirable to store a SRID integer in each geometry blob, when 99% of systems and formats (except PostGIS) only deal with it on a column basis. PostGIS's integer SRIDs are also not guaranteed to be compatible across systems either (not even across databases, you can alter the SPATIAL_REF_SYS table yourself to add your own if you want), so you need to store something more precise out-of-band anyway. If you really want mixed integer SRIDs per-row, why not just store it in a separate column? Every spatial function is oblivious to the projection anyway, except `ST_Transform`, but it has overloads to both take the target/source CRS by argument or extracting it from the column type parameter.

I think the only thing we could really do to improve the situation right now is to automatically try to resolve the projection associated with a column by joining the `geometry_columns` view with the `SPATIAL_REF_SYS` table on import/export to DucKDB. But doing this on export (from duckdb to postgis) is also a bit tricky without PROJ or the spatial extension as you need to determine CRS equivalence. It's been on my mind to try to break out our PROJ dependency from spatial into a small lightweight extension that could provide e.g. scalar functions to do these sort of things - but that's also tricky as long as spatial uses GDAL since we need PROJ compiled-in.

**Comment by tjwebb** [comment]
+1

**Comment by TomvanTilburg-geodan** [comment]
For inspiration: linking a recent post from Paul Ramsey related to the topic (on empty geometries and how to store geometry in general):
https://blog.cleverelephant.ca/2025/02/wkb-empty.html

**Comment by palmerj** [comment]
Is there a plan to add this feature to the roadmap?
